haiku/_src/data_structures.py

```diff
@@ -92,12 +92,14 @@ def __repr__(self):
 # TODO(lenamartens) Deprecate type
 class frozendict(Mapping[K, V]):  # pylint: disable=invalid-name
   """Immutable mapping from keys to values."""
+  _isinitialized = False
```

Comment:
It would be better to use `__slots__` to handle this:

```python
class frozendict(Mapping[K, V]):  # pylint: disable=invalid-name
  """Immutable mapping from keys to values."""
  __slots__ = ("_storage", "_keys", "_hash")
```

The benefits of slots over setattr:

  1. Python standard way to handle this.
  2. Minor performance/space improvement for instances.
  3. IDEs understand slots and know which properties can and cannot be set.
  4. Tools like pytype/mypy can also detect violations using static analysis.

Reply:
You're right, that's much better. Thanks @tomhennigan